src/main/java/io/confluent/connect/hdfs/orc/OrcFileReader.java

```diff
@@ -0,0 +1,118 @@
+import org.apache.hadoop.hive.serde2.objectinspector.primitive.PrimitiveObjectInspectorUtils;
+import org.apache.hadoop.hive.serde2.objectinspector.primitive.PrimitiveObjectInspectorUtils.PrimitiveTypeEntry;
+import org.apache.kafka.connect.data.ConnectSchema;
+import org.apache.kafka.connect.data.Date;
+import org.apache.kafka.connect.data.Schema;
+import org.apache.kafka.connect.data.SchemaBuilder;
+import org.apache.kafka.connect.data.Timestamp;
+import org.apache.kafka.connect.errors.ConnectException;
+import org.apache.kafka.connect.errors.DataException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.Iterator;
+
+public class OrcFileReader implements SchemaFileReader<HdfsSinkConnectorConfig, Path> {
+
+  private static final Logger log = LoggerFactory.getLogger(OrcFileReader.class);
+
+  @Override
+  public Schema getSchema(HdfsSinkConnectorConfig conf, Path path) {
+    try {
+      log.info("Opening ORC record reader for: {}", path);
+
+      ReaderOptions readerOptions = new ReaderOptions(conf.getHadoopConfiguration());
+      Reader reader = OrcFile.createReader(path, readerOptions);
+
+      if (reader.getObjectInspector().getCategory() == ObjectInspector.Category.STRUCT) {
+        SchemaBuilder schemaBuilder = SchemaBuilder.struct().name("record").version(1);
+        StructObjectInspector objectInspector = (StructObjectInspector) reader.getObjectInspector();
+
+        for (StructField schema : objectInspector.getAllStructFieldRefs()) {
+          ObjectInspector fieldObjectInspector = schema.getFieldObjectInspector();
+          String typeName = fieldObjectInspector.getTypeName();
+          Schema.Type schemaType;
+
+          switch (fieldObjectInspector.getCategory()) {
+            case PRIMITIVE:
+              PrimitiveTypeEntry typeEntry = PrimitiveObjectInspectorUtils
+                  .getTypeEntryFromTypeName(typeName);
+              if (java.sql.Date.class.isAssignableFrom(typeEntry.primitiveJavaClass)) {
+                schemaType = Date.SCHEMA.type();
+              } else if (java.sql.Timestamp.class.isAssignableFrom(typeEntry.primitiveJavaClass)) {
+                schemaType = Timestamp.SCHEMA.type();
+              } else {
+                schemaType = ConnectSchema.schemaType(typeEntry.primitiveJavaClass);
+              }
+              break;
+            case LIST:
+              schemaType = Schema.Type.ARRAY;
+              break;
+            case MAP:
+              schemaType = Schema.Type.MAP;
+              break;
+            default:
+              throw new DataException("Unknown type " + fieldObjectInspector.getCategory().name());
+          }
+
+          schemaBuilder.field(schema.getFieldName(), SchemaBuilder.type(schemaType).build());
+        }
+
+        return schemaBuilder.build();
+      } else {
+        throw new ConnectException(
+            "Top level type must be of type STRUCT, but was "
```

Comment:
Is this necessary?
Why doesn't it support primitive root types?

Reply:
ORC is a column file format so `Struct` is required